README.md
# auction-house

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-09-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-09-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/auction-house/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/auction-house/compare/0.1.0...0.1.1


test/auction_house/core_test.clj
(ns auction-house.core-test
  (:require [clojure.test :refer :all]
            [auction-house.core :refer :all]))

(deftest exercice-example-test
  (testing "auction with 100 euros as reserve price and:
    Bob bidding 110, 130,
    Eva no bidding,
    Marc bidding 125
    Ben bidding 105, 115, 90
    David bidding 132, 135, 140
    should end with David as winner with a price of 130 euro"
    (is (= "Winner is :david with a bid of 130" (winning-auction 100 {:bob   #{110 130}
                                                                      :eva   #{}
                                                                      :marc  #{125}
                                                                      :ben   #{105 115 90}
                                                                      :david #{132 135 140}})))))
(deftest auction-test-suite
  (testing "empty auction with 10 Euros as reserve price and no bidder should end with no winner"
    (is (= "No winner for this auction" (winning-auction 10 {}))))

  (testing "empty auction with 0 euros as reserve price and no bidder should end with no winner"
    (is (= "No winner for this auction" (winning-auction 0 {}))))

  (testing "empty auction with 0 euros as reserve price and no auction bids should end with no winner"
    (is (= "No winner for this auction" (winning-auction 0  nil))))

  (testing "auction with 0 euros as reserve price and Bob as single bidder with 0 should end with Bob as winner"
    (is (= "Winner is :bob with a bid of 0" (winning-auction 0 {:bob #{0}}))))

  (testing "auction with 0 euros as reserve price and Bob as single bidder with 1 should end with Bob as winner"
    (is (= "Winner is :bob with a bid of 0" (winning-auction 0 {:bob #{1}}))))

  (testing "auction with 0 euros as reserve price and Bob as single bidder with negative amount value should end with no winner"
    (is (= "No winner for this auction" (winning-auction 0 {:bob #{-1}}))))

  (testing "empty auction with 0 euros as reserve price and 2 bidders with no bids should end with no winner"
    (is (= "No winner for this auction" (winning-auction 0 {:bob #{} :tommy #{}}))))

  (testing "empty auction with 0 euros as reserve price and 2 bidder with no bids should end with no winner"
    (is (= "No winner for this auction" (winning-auction 0 {:bob nil :tommy nil}))))

  (testing "auction with 0 euros as reserve price and 2 bidders with a 0 amount each should end with no winner"
    (is (= "No winner for this auction" (winning-auction 0 {:bob #{0} :tom #{0}}))))

  (testing "auction with 0 euros as reserve price and 2 bidders with amount of 1 each should end with no winner"
    (is (= "No winner for this auction" (winning-auction 0 {:bob #{1} :tom #{1}}))))

  (testing "auction with 1 euros as reserve price and Bob bidding 1, and tom bidding 2, should end with Tom as winner with a price of 1 euro"
    (is (= "Winner is :tom with a bid of 1" (winning-auction 1 {:bob #{1} :tom #{2}}))))

  (testing "auction with 2 euros as reserve price and Bob bidding 1 and David bidding 2, should end with David as winner with a price of 2 euro"
    (is (= "Winner is :david with a bid of 2" (winning-auction 2 {:bob #{1} :david #{2}}))))

  (testing "auction with 20 euros as reserve price and Bob bidding 21, 22 and tom bidding 1,9 should end with Bob as winner with the reserved price"
    (is (= "Winner is :bob with a bid of 20" (winning-auction 20 {:bob #{21 22} :tom #{19}}))))

  (testing "auction with 19 euros as reserve price and Bob bidding 21, 22, 23 and tom bidding 20, should end with Bob as winner with the last winning bid (20)"
    (is (= "Winner is :bob with a bid of 20" (winning-auction 19 {:bob #{21 22 23} :tom #{20}}))))

  (testing "unordered auction with 19 euros as reserve price and Bob bidding 22, 21, 23 and tom bidding 21, 20, 19, should end with Bob as winner with the last winning bid (21)"
    (is (= "Winner is :bob with a bid of 21" (winning-auction 19 {:bob #{22 21 23} :tom #{21 20 19}})))))


src/auction_house/core.clj
(ns ^{:doc "The Auction house program."
      :author "Benjamin Martinez Ibarra"}
  auction-house.core
  (:gen-class))

(declare find-winning-bid
         map-to-bids
         amount bidder
         to-bids
         sort-by-amount-desc
         same-bidder?
         winning-bidder-amount
         by-bid-amount
         minimum-winning-amount)

;Bid type
(defrecord Bid
  [bidder bid-amount])

(defn winning-auction
  "Main controller of the auction. Will compare each bid until finding a winner.
  Returns the winner of the auction with the bidding amount he will have to pay.
  A winner has the best bid amount for the auction but may not pay that amount.
  A winner will pay the last winning bid amount involving another bidder or the start price if not superior.
  May return 'no winner' message if no bidder are present and/or none can afford the start price at least."
  [start-price all-bids]
  (let [winning-bid (-> all-bids
                        (map-to-bids)
                        (find-winning-bid start-price))]
    (if (= :no-winner winning-bid)
      "No winner for this auction"
      (let [winning-amount (amount winning-bid)
            winner (bidder winning-bid)]
        (str "Winner is " winner " with a bid of " winning-amount)))))

(defn- map-to-bids
  "Transform a map where each entry represents a bidder and his bids amounts into a sequence of Bids ([bidder bid-amount] pairs)."
  [bidder-amounts-map]
  (reduce
    (fn [all-bids [bidder bid-amounts]]
      (into all-bids (to-bids bidder bid-amounts)))
    (vector)
    (seq bidder-amounts-map)))

(defn- to-bids
  "Transforms a bidder and its bids amounts into a sequence of Bids.
  A Bid is composed of a bidder and a bid-amount."
  [bidder bids-amounts]
  (reduce
    (fn [bids bid-amount]
      (conj bids (->Bid bidder bid-amount)))
    (vector)
    bids-amounts))

(defn find-winning-bid
  "Finds the winning bid among a collection of bids.
   May return :no-winner keyword if none of the 2 best bidders have a bidding amount >= start price
   or both of the two greatest bidders have same max bid amounts.
   Otherwise, iterate through the bids and find the combination of the best bidder whose bid amount is the greatest among all the bid
   and the last winning bid amount value whenever it comes from a different bidder."
  ([bids start-price]
   (let [sorted-bids (sort-by-amount-desc bids)
         top-bid (first sorted-bids)
         next-bids (rest sorted-bids)]
     (if (or (nil? top-bid)
             (< (amount top-bid) start-price)
             (and (= 2 (count sorted-bids))
                  (= (amount top-bid) (amount (first next-bids)))))
       :no-winner
       (find-winning-bid top-bid next-bids start-price))))
  ([reference-bid bids start-price]
   (let [next-bid (first bids)]
     (if (not (same-bidder? reference-bid next-bid))
       (winning-bidder-amount reference-bid next-bid start-price)
       (recur next-bid (rest bids) start-price)))))

(defn- sort-by-amount-desc
  "Sorts a list of bids by their bid-amount."
  [bids]
  (sort by-bid-amount bids))

(defn by-bid-amount
  "Simple comparator between 2 bids, comparing their amount decremently."
  [^Bid bid1 ^Bid bid2]
  (compare (amount bid2) (amount bid1)))

(defn- amount
  "Retrieve the amount part of a bid"
  [^Bid bid]
  (or (:bid-amount bid) 0))

(defn- same-bidder?
  "Returns true if both bids belong to the same bidder.False otherwise"
  [^Bid bid1 ^Bid bid2]
  (= (bidder bid1) (bidder bid2)))

(defn- bidder
  "Retrieve the bidder part of a bid"
  [^Bid bid]
  (:bidder bid))

(defn- winning-bidder-amount
  "Compare two bids and returns a combination of:
    - the bidder whose bid amount is the greatest of the two bids.
    - the minimum winning amount of the 2 bids or start-price if the amount is lower."
  [first-bid second-bid start-price]
  (let [first-bid-amount (amount first-bid)
        second-bid-amount (amount second-bid)]
    (cond
      (= first-bid-amount second-bid-amount) (->Bid (bidder first-bid) first-bid-amount)
      (> first-bid-amount second-bid-amount) (->Bid (bidder first-bid) (minimum-winning-amount second-bid-amount start-price))
      :else (->Bid (bidder second-bid) (minimum-winning-amount second-bid-amount start-price)))))

(defn- minimum-winning-amount
  "Retrieve the lowest amount between a bid amount and a start price.
   That amount can't be lower than the reference price."
  [last-winning-bid-amount start-price]
  (if (>= last-winning-bid-amount start-price)
    last-winning-bid-amount
    start-price))

(defn -main
  "Retrieving the winner of an auction."
  [& _]
  (let [bids {:bob   #{110 130}
              :eva   #{}
              :marc  #{125}
              :ben   #{105 115 90}
              :david #{132 135 140}}
        winning-bid (winning-auction 100 bids)]
    (println "auction: ")
    (println bids "\n")
    (println winning-bid)))
